### PR TITLE
CORTX-33454 copy object conditions

### PR DESCRIPTION
**Pull request by archanabpatil**
# Problem Statement
- Added tests for conditional parameters :https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1079640340/TDS+F-20A+S3+Copy+Object+AP

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/s3/test_copy_object.py [hunk 1]
```diff
@@ -2257,3 +2262,580 @@ def test_22299(self):
             LOGGER.info(err.message)
             assert_utils.assert_in(errmsg.ACCESS_DENIED_ERR_KEY,err.message, err)
         LOGGER.info("ENDED: Use bucket policy to validate copy object with applied ACL.")
+
+    @pytest.mark.parallel
+    @pytest.mark.s3_ops
+    @pytest.mark.s3_object_copy
+    @pytest.mark.tags("TEST-")
```
Comment (ketanarlulkar) [inline tests/s3/test_copy_object.py:2269]: Add test ID tags to all tests

**Comment by archanabpatil** [comment]
This is just TA. TE is not done , its planned for next sprint

File: tests/s3/test_copy_object.py (hunk 1 above)
Comment (archanabpatil) [reply tests/s3/test_copy_object.py]: Will do once I create Jira tickets for these tests  . Waiting on dev to finish the review

Comment (archanabpatil) [reply tests/s3/test_copy_object.py]: Done @ketanarlulkar

File: tests/s3/test_copy_object.py [hunk 2]
```diff
@@ -2257,3 +2261,809 @@ def test_22299(self):
...
+    @pytest.mark.parallel
+    @pytest.mark.s3_ops
+    @pytest.mark.s3_object_copy
+    @pytest.mark.tags("TEST-44750")
+    @CTFailOn(error_handler)
+    def test_44750(self):
+        """
+        Copy object Conditional params: All true for simple object copy"""
+        LOGGER.info(
+            "STARTED: Copy simple object with all conditional headers as true")
```
Comment (ketanarlulkar) [inline tests/s3/test_copy_object.py:2274]: Do not break statements unless it crosses 100 characters.

File: tests/s3/test_copy_object.py [hunk 3]
```diff
@@ -2257,3 +2261,809 @@ def test_22299(self):
...
+        except CTException as error:
+            assert_utils.assert_in(errmsg.RGW_ERR_CPY_IF_MODIFIED_SINCE_FALSE, error.message, error)
+        LOGGER.info("Step 6: Copy object to different bucket with condition "                             
+                    "x-amz-copy-source-if-unmodified-since (False)")
+        try:
+            status, response = self.s3_obj.copy_object(
+                self.bucket_name1, self.object_name1, self.bucket_name2,
+                self.object_name2 + "_iums", CopySourceIfUnmodifiedSince=pre_date)
+            assert_utils.assert_false(status, response)
+        except CTException as error:
```
Comment (ketanarlulkar) [inline tests/s3/test_copy_object.py:2444]: The try except can go in helper function with errmsg as one of the parameter

File: tests/s3/test_copy_object.py (hunk 2 above)
Comment (archanabpatil) [reply tests/s3/test_copy_object.py:2274]: yes

File: tests/s3/test_copy_object.py [hunk 4]
```diff
@@ -2257,3 +2261,809 @@ def test_22299(self):
...
+    @pytest.mark.s3_object_copy
+    @pytest.mark.tags("TEST-44752")
+    @CTFailOn(error_handler)
+    def test_44752(self):
+        """
+        Copy object Conditional params: All true for multipart object copy"""
+        LOGGER.info(
+            "STARTED: Copy multipart object with all conditional headers as true")
+        _, response = self.s3_obj.create_bucket(self.bucket_name1)
+        assert_utils.assert_true(response[0], response[1])
```
[the remark this answers is not in the document]
Comment (archanabpatil) [reply tests/s3/test_copy_object.py:2335]: yup

File: tests/s3/test_copy_object.py (hunk 3 above)
Comment (archanabpatil) [reply tests/s3/test_copy_object.py:2444]: I can do that during TE phase as part of RVF

File: commons/error_messages.py [hunk 5]
```diff
@@ -52,6 +52,14 @@
                                "operation: This copy request is illegal because it is trying to "
                                "copy an object to itself without changing the object's metadata, "
                                "storage class, website redirect location or encryption attributes.")
+RGW_ERR_CPY_IF_MATCH_FALSE  = ("An error occurred (PreconditionFailed) when calling the "
+                               "CopyObject operation: Unknown")
+RGW_ERR_CPY_IF_NONE_MATCH_FALSE  = ("An error occurred (304) when calling the CopyObject "
+                                     "operation : Not Modified")
```
Comment (apoorva-rao) [inline commons/error_messages.py:58]: Please fix indentation in code block 58 - 62

File: tests/s3/test_copy_object.py [hunk 6]
```diff
@@ -2257,3 +2261,809 @@ def test_22299(self):
...
+        LOGGER.info("Step 2: Copy object to different bucket")
+        status, response = self.s3_obj.copy_object(self.bucket_name1, self.object_name1,
+                                                   self.bucket_name2, self.object_name2)
+        assert_utils.assert_true(status, response)
+        obj_list.append(self.object_name2)
+        LOGGER.info("Step 3: Copy object to different bucket with condition "
+                    "x-amz-copy-source-if-match (True)")
+        status, response = self.s3_obj.copy_object(self.bucket_name1, self.object_name1,
+                                                   self.bucket_name2, self.object_name2 + "_im",
+                CopySourceIfMatch=response[1]["ETag"])
```
Comment (apoorva-rao) [inline tests/s3/test_copy_object.py:2293]: Please fix indentation in copy_object calls across the file, wherever applicable

**Review by apoorva-rao — CHANGES_REQUESTED** [review]
@archanabpatil some indentation fixes needed, please see the inline comments

File: commons/error_messages.py (hunk 5 above)
Comment (archanabpatil) [reply commons/error_messages.py:58]: yes

File: tests/s3/test_copy_object.py (hunk 6 above)
Comment (archanabpatil) [reply tests/s3/test_copy_object.py]: its available in latest commit

File: tests/s3/test_copy_object.py [hunk 7]
```diff
@@ -2257,3 +2261,808 @@ def test_22299(self):
...
+        status, resp = self.s3_obj.copy_object(self.bucket_name1, self.object_name1,
+                                               self.bucket_name2, self.object_name2+"_im",
+                                               CopySourceIfMatch=etag)
+        assert_utils.assert_true(status, resp)
+        obj_list.append(self.object_name2 + "_im")
+        LOGGER.info("Step 4: Copy object to different bucket with condition"
+                    "x-amz-copy-source-if-none-match (True)")
+        status, response = self.s3_obj.copy_object(self.bucket_name1, self.object_name1,
+                                                   self.bucket_name2, self.object_name2+"_inm",
+                                                    CopySourceIfNoneMatch=etag+"_")
```
Comment (AshwiniS-Borse) [inline tests/s3/test_copy_object.py:2302]: Indentation fix required

File: tests/s3/test_copy_object.py [hunk 8]
```diff
@@ -2257,3 +2261,808 @@ def test_22299(self):
...
+                    "x-amz-copy-source-if-modified-since (False)")
+        date1 = response["CopyObjectResult"]["LastModified"]
+        date1 = datetime.strptime(date1, '%b %d %Y')
+        pre_date = date1 - timedelta(days=1)
+        post_date = date1 + timedelta(days=1)
+        try:
+            status, response = self.s3_obj.copy_object(self.bucket_name1, self.object_name1,
+                                                       self.bucket_name2,
+                                                       self.object_name2+"_ims",
+                                                        CopySourceIfModifiedSince=post_date)
```
Comment (AshwiniS-Borse) [inline tests/s3/test_copy_object.py:2437]: Indentation

**Review by apoorva-rao — APPROVED** [review]

**Review by AshwiniS-Borse — CHANGES_REQUESTED** [review]

**Review by ketanarlulkar — APPROVED** [review]

**Review by AshwiniS-Borse — APPROVED** [review]

File: tests/s3/test_copy_object.py (hunk 7 above)
Comment (archanabpatil) [reply tests/s3/test_copy_object.py:2302]: Will do as part of RVF in next sprint. I anyways have to do code optimization as well then

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**